### PR TITLE
Extract some methods from ci lazy methods

### PR DESCRIPTION
**Pull request by romainbrenguier**
This extracts some functions form the ci_lazy_methods operator() whose definition is too long.
I'm not sure about the naming of the extracted function, so please suggest better names if you can.

### CODE REVIEW

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 1]
```diff
@@ -297,6 +271,40 @@ bool ci_lazy_methodst::operator()(
   return false;
 }
 
+std::unordered_set<irep_idt>
+ci_lazy_methodst::reachable_from_main_function(const symbol_tablet &symbol_table)
```
Comment (thk123) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp]: You might want to document here the fact that is there is no main method it gets all the methods from the main class (and if no main class, all the menthods!)

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 2]
```diff
@@ -255,6 +255,50 @@ bool ci_lazy_methodst::operator()(
   return false;
 }
 
+// Look for virtual callsites with no candidate targets. If we have
+// invokevirtual A.f and we don't believe either A or any of its children
+// may exist, we assume specifically A is somehow instantiated. Note this
+// may result in an abstract class being classified as instantiated, which
+// stands in for some unknown concrete subclass: in this case the called
+// method will be a stub.
+void ci_lazy_methodst::insert_virtual_methods(
```
Comment (thk123) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp]: Possibly GH being weird - but I can't see this method being used anywhere in this PR (or where it is being extracted from?)

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 3]
```diff
@@ -218,11 +218,27 @@ bool ci_lazy_methodst::operator()(
...
 
+  debug() << "CI lazy methods: removed "
+          << symbol_table.symbols.size() - keep_symbols.symbols.size()
+          << " unreachable methods and globals" << eom;
+
+  symbol_table.swap(keep_symbols);
+  return false;
+}
+
+symbol_tablet ci_lazy_methodst::filter_symbol_table(
```
Comment (thk123) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp]: Document please - I'd expect a filter method to take a predicate

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 4]
```diff
@@ -98,34 +98,8 @@ bool ci_lazy_methodst::operator()(
...
-          "java::" + id2string(class_name) + "." + id2string(method.name)
-            + ":" + id2string(method.descriptor);
-        methods_to_convert_later.insert(methodid);
-      }
-    }
-  }
-  else
-    methods_to_convert_later.insert(main_function.main_function.name);
+  std::unordered_set<irep_idt> methods_to_convert_later =
+    reachable_from_main_function(symbol_table);
```
Comment (smowton) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp]: Rename `entry_point_methods`

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 5]
```diff
@@ -287,14 +261,126 @@ bool ci_lazy_methodst::operator()(
...
-  debug() << "CI lazy methods: removed "
-          << symbol_table.symbols.size() - keep_symbols.symbols.size()
-          << " unreachable methods and globals" << eom;
+// Look for virtual callsites with no candidate targets. If we have
+// invokevirtual A.f and we don't believe either A or any of its children
+// may exist, we assume specifically A is somehow instantiated. Note this
+// may result in an abstract class being classified as instantiated, which
+// stands in for some unknown concrete subclass: in this case the called
+// method will be a stub.
+void ci_lazy_methodst::insert_virtual_methods(
```
Comment (smowton) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp]: Rename `handle_virtual_methods_with_no_callees`

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 6]
```diff
@@ -287,14 +261,126 @@ bool ci_lazy_methodst::operator()(
...
+    const irep_idt method_name = get_virtual_method_target(
+      instantiated_classes, call_basename, call_class, symbol_table);
+    CHECK_RETURN(!method_name.empty());
+
+    // Add what it returns to methods_to_convert_later
+    methods_to_convert_later.insert(method_name);
+  }
+}
+
+ci_lazy_methodst::convert_method_resultt ci_lazy_methodst::convert_method(
```
Comment (smowton) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp]: Rename `convert_and_analyse_method`

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp (hunk 2 above)
Comment (romainbrenguier) [reply jbmc/src/java_bytecode/ci_lazy_methods.cpp]: Not github's fault, either I lost a commit or the refactoring forgot to replace the old code by a call to this function.
I corrected that.

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 7]
```diff
@@ -255,6 +223,52 @@ bool ci_lazy_methodst::operator()(
...
+/// stands in for some unknown concrete subclass: in this case the called
+/// method will be a stub.
+/// \return whether a new class was encountered
+bool ci_lazy_methodst::handle_virtual_methods_with_no_callees(
+  std::unordered_set<irep_idt> &methods_to_convert_later,
+  std::unordered_set<irep_idt> &instantiated_classes,
+  const std::unordered_set<exprt, irep_hash> &virtual_function_calls,
+  symbol_tablet &symbol_table)
+{
+  bool any_new_classes;
```
Comment (thk123) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp]: Give default value! Currently if `virtual_function_calls` is empty it is UB what this returns

**Review by thk123 — CHANGES_REQUESTED** [review]

**Review by thk123 — APPROVED** [review]

**Comment by thk123** [comment]
TG PR is currently failing

**Comment by romainbrenguier** [comment]
@smowton I renamed the functions according to your suggestions

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 8]
```diff
@@ -297,6 +223,134 @@ bool ci_lazy_methodst::operator()(
...
+      instantiated_classes, call_basename, call_class, symbol_table);
+    CHECK_RETURN(!method_name.empty());
+
+    // Add what it returns to methods_to_convert_later
+    methods_to_convert_later.insert(method_name);
+  }
+  return any_new_classes;
+}
+
+ci_lazy_methodst::convert_method_resultt
```
Comment (smowton) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp]: Needs a doc comment

**Comment by romainbrenguier** [comment]
@smowton I added some documentation, can you check it's correct?

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 9]
```diff
@@ -152,46 +126,30 @@ bool ci_lazy_methodst::operator()(
...
+        std::unordered_set<irep_idt> methods_to_convert;
+        std::swap(methods_to_convert, methods_to_convert_later);
+        for(const auto &mname : methods_to_convert)
         {
-          class_initializer_seen = true;
-          irep_idt initializer_signature =
-            get_java_class_literal_initializer_signature();
-          if(symbol_table.has_symbol(initializer_signature))
-            methods_to_convert_later.insert(initializer_signature);
+          const auto convertion_result = convert_and_analyze_method(
```
Comment (smowton) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp]: sp `convertion` -> `conversion`

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 10]
```diff
@@ -297,6 +223,143 @@ bool ci_lazy_methodst::operator()(
...
+///     * class_initializer_seen which is true if the class_initializer_seen
+///       argument was false and the class_model is referenced in
+///       the body of the method
+///     * new_method_seen if the method was not converted before and was
+///       converted successfully here
+ci_lazy_methodst::convert_method_resultt
+ci_lazy_methodst::convert_and_analyze_method(
+  const method_convertert &method_converter,
+  std::unordered_set<irep_idt> &methods_already_populated,
+  const bool class_initializer_seen,
```
Comment (smowton) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp]: Rename `class_initializer_already_seen`

**Review by smowton — APPROVED** [review]
Couple of nitpicks otherwise  lgtm

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 610bbdf0507cf76c6d1b6b6434fb7f75766e7388).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 2ffbd24339c6ba6ae833e943cae2859f0c1c0507).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: c4aadabc22cdb24e8eb116690e718cd56dd50d42).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76499418

**Pull request merged**